pkg/deploy/depoy.go

```diff
@@ -0,0 +1,247 @@
+	if len(os.Getenv("AZURE_RP_PREDEPLOY")) != 0 {
+		var data []byte
+		data, err = DeployRpProductionNsgJsonBytes()
+		if err != nil {
+			return
+		}
+
+		var azuretemplate map[string]interface{}
+		err = json.Unmarshal(data, &azuretemplate)
+		if err != nil {
+			return
+		}
+
+		log.Infof("deploying nsg and managedIdentity to %s", d.resourceGroupName)
+		err = d.deployments.CreateOrUpdateAndWait(ctx, d.resourceGroupName, "rp-production-nsg", azresources.Deployment{
+			Properties: &azresources.DeploymentProperties{
+				Template: azuretemplate,
+				Mode:     azresources.Incremental,
+			},
+		})
+		if err != nil {
+			return
+		}
+	}
+	deployment, err := d.deployments.Get(ctx, d.resourceGroupName, "rp-production-nsg")
+	if err != nil {
+		return
+	}
+	return deployment.Properties.Outputs.(map[string]interface{})["rpServicePrincipalId"].(map[string]interface{})["value"].(string), nil
+}
+
+func (d *deployer) Deploy(ctx context.Context, log *logrus.Entry, rpServicePrincipalID string) error {
+	data, err := DeployRpProductionJsonBytes()
+	if err != nil {
+		return err
+	}
+
+	var azuretemplate map[string]interface{}
+	err = json.Unmarshal(data, &azuretemplate)
+	if err != nil {
+		return err
+	}
+
+	parameters, err := getParameters()
+	if err != nil {
+		return err
+	}
+	parameters["vmssName"].(map[string]interface{})["value"] = d.version
+	parameters["rpServicePrincipalId"].(map[string]interface{})["value"] = rpServicePrincipalID
+	// azure enforce ^[a-z][a-z0-9-]{1,61}[a-z0-9]$. for DomainNameLabel.
+	// gitCommit do not comply as it might start with a number
+	parameters["vmssDomainNameLabel"].(map[string]interface{})["value"] = "vmss-" + d.version
+	d.parameters = parameters
+
+	log.Infof("deploying rp version %s to %s", d.version, d.resourceGroupName)
+	return d.deployments.CreateOrUpdateAndWait(ctx, d.resourceGroupName, "rp-production-"+d.version, azresources.Deployment{
+		Properties: &azresources.DeploymentProperties{
+			Template:   azuretemplate,
+			Mode:       azresources.Incremental,
+			Parameters: parameters,
+		},
+	})
+}
+
+func (d *deployer) Upgrade(ctx context.Context, log *logrus.Entry) error {
+	scaleSets, err := d.vmss.List(ctx, d.resourceGroupName)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("checking new %s RP health", d.version)
+	err = d.checkRPReadiness(ctx, log, "rp-vmss-"+d.version)
+	if err != nil {
+		return err
+	}
+
+	log.Info("retire old RP")
+	err = d.retireOldVMSS(ctx, log, scaleSets)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *deployer) checkRPReadiness(ctx context.Context, log *logrus.Entry, vmssName string) error {
+	scaleSetsVMs, err := d.vmssvms.ListVMs(ctx, d.resourceGroupName, vmssName, "", "", "")
+	if err != nil {
+		return err
+	}
+
+	// construct readiness tracking map with FQDN for checking
+	var urlPool []string
+	for _, vm := range scaleSetsVMs {
+		url := fmt.Sprintf("https://vm%s.%s.%s.cloudapp.azure.com/healthz/ready", *vm.InstanceID, "vmss-"+d.version, d.location)
```

Comment:
`"vmss-"+d.version`: can you actually read that from some field of the `vm` object instead of assuming?